### ISSUE TITLE
Handling dual controllers vs left and right joysticks

### ISSUE THREAD

**Issue by rpavlik** [OSVR/OSVR-Core#283]
We've got a bit of a conundrum regarding tree structuring for controllers.

We've used the razer hydra controllers as `/controller/left` and `/controller/right`. To match the rest of the practice (like `/me/hands/left`), they should probably be `/controllers/left` and `/controllers/right`.

However, then, where do we map gamepads (which are currently just mapped to `/controller`?)

Additionally, two-handed controllers like an xbox controller often have two joysticks. To follow the pattern, those would be `joystick/left` and `joystick/right` with axes beneath them, but we've also indicated a tendency to have `joystick` be a logical collection of its axes and optionally button, which is presumably at least one reason why current xbox descriptors refer to `left/joystick/x`, `left/joystick/y`, etc. even though that would conflate it with the hydra's two hands.

**Comment by russell-taylor** [comment]
Yep, time to sit down and work out the One True Way.  Not that others can't deviate from it, but at least people would know the default is.

This seems to be a separate hierarchy that is independent of the spatial hierarchy, and it has the interesting feature that it is not a tree: multiple nodes in the hierarchy can point to the same location.  There is some conflation going on, with /me/hands/left potentially including both spatial information and other information (buttons, analogs, dials, ...).  This is naturally going to cause some confusion when using a two-handed controller, because there should be no spatial information and both hands alias back onto the same device.

I think we're trying to factor this space separately according to device description and according to function in a manner that is truly orthogonal.  We also want to pick different names for the objects in the two factors, so we don't end up using the plural and singular of the same name (hand, controller) to mean two different things.

Here's one potential approach:

(1) Make a device-specific descriptive hierarchy that is hardware specific and not based on function or location.  Always consider a situation where multiple devices may be used by the system (to enable multiple users, or people doing crazy things like duct-taping to the floor or the HMD or their wrist).  I think that "joystick" is probably too specific for the top level of this factorization and would tend to go with controller0/leftThumbJoystick/x (an analog) and controller0/start (a button).  If we don't go with Hungarian notation, it may be ambiguous to the user what kind of object each thing is: controller0/leftThumbPad/left would be a button but people might think of that device as a hi-hat (and it may be implemented both ways), but doing controller0/leftThumbPad/leftButton seems extra wordy; I might go with extra wordy, but that puts Analog and Button on the right side of every non-tracker-object.  controller0/leftThumbPad/x or controller/leftThumbPad/xAnalog would be the hi-hat interface, and I guess that is different enough that maybe we don't need Hungarian.

(2) Make a function-specific alias overlay.  This may require the ability to at least invert, and potentially also scale and offset.  me/flyForward would map to a negated controller0/leftThumbJoystick/y in the Visual Common Ground application, and me/flyRight would map to controller0/LeftThumbJoystick/x.  me/quit would map to controller0/start (which is underneath an alias to controller0/buttons/8 (maybe not that exact number, but you get the idea).

**Comment by russell-taylor** [comment]
I've recently come across a similar issue, which you may want to keep here or move to another issue.  It is possible to construct a config file that has no /me/head defined in it.  Visual Common Ground was not checking the return on the read for this, and thereby getting bogus values that broke the transforms.

This ties into the internal OSVR hierarchy and so may be a discussion point for version 2.0, but some potential solutions are: (1) Define some spaces as being required and (a) fail to load the server if they are not defined or (b) always construct default ones if they are not defined.  (2) Add checking for the existence of these spaces into the example programs so that people know that this is something that can fail.  Of course, optional spaces like hands need to be checked for (and are), but it was surprising as a user to find that /me/head did not exist.

**Comment by rpavlik** [comment]
The server should probably warn on app connect if /me/head doesn't exist - please file that separately.
